amy/extrequests/forms.py

```diff
@@ -440,6 +453,31 @@ def institution_label_from_instance(obj):
         essentially works just like `Model.__str__`."""
         return "{}".format(obj.fullname)
 
+    @feature_flag_enabled("ENFORCE_MEMBER_CODES")
+    def clean_membership_info(self, request: HttpRequest) -> dict:
+        errors = dict()
+        affiliation = self.cleaned_data.get("member_affiliation")  # yes/no/unsure
+        code = self.cleaned_data.get("member_code", "")
+        if affiliation in ["yes", "unsure"] and code:
+            # ensure that code belongs to a membership
+            try:
+                Membership.objects.get(registration_code=code)
+            except Membership.DoesNotExist:
+                errors["member_code"] = ValidationError(
+                    "This code is invalid. "
+                    "Please contact your Member Affiliate to verify your code."
+                )
+        elif affiliation == "yes" and not code:
+            errors["member_code"] = ValidationError(
+                "This field is required if you selected 'Yes' above."
+            )
+        elif affiliation == "no" and code:
+            errors["member_code"] = ValidationError(
+                "This field must be empty if you selected 'No' above."
+            )
+
+        return errors
```

Comment:
I don't think we have a place in AMY where we do "soft validation". Here are my three ideas how to implement it. It's definitely not easy.

**Idea 1: code validation as background process in the browser**

Once the code is filled out by the user, we'd make a request to the API to see if there's a membership matching criteria.

With JS we'd dynamically show warnings in the form for the user, based on the responses from the server.

This potentially exposes the API to third parties as users don't have to be logged in in these forms, so some throttling should be implemented.


**Idea 2: dynamic field requesting user to confirm the message popup**

After user provides the code and there's something wrong with it, we amend the form by adding a new required checkbox field next to that member code.

This is clever because it uses Django default behavior to "break" form validation, but at the same time it's easy for the user to correct the form without changing member code - thus soft validation.

The downside is that altering a form like that is not an easy task.

**Idea 3: remembering form state between requests**

This idea uses Django session storage to remember if user say the warning message.

When first attempt on validating the form throws the error like in your current implementation, the server should also store some variable in the session.

This variable, when present in the session, would then change the member code validation and prevent throwing the same error twice. Finally the variable should be removed from the server session.

Reply:
Thanks for all these - I thought of idea 2 myself as well. I think idea 2 would be best as then it makes it easy to flag to the admins when the code needs to be double-checked.

Reply:
Created #2535 from this comment